dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/trackedentity/DefaultTrackedEntityChangeLogService.java

```diff
@@ -0,0 +1,133 @@
+import org.hisp.dhis.trackedentity.TrackedEntity;
+import org.hisp.dhis.trackedentity.TrackedEntityAttribute;
+import org.hisp.dhis.trackedentity.TrackedEntityAttributeService;
+import org.hisp.dhis.trackedentity.TrackedEntityProgramOwnerService;
+import org.hisp.dhis.trackedentity.TrackedEntityService;
+import org.hisp.dhis.trackedentity.TrackerAccessManager;
+import org.hisp.dhis.tracker.export.Page;
+import org.hisp.dhis.tracker.export.PageParams;
+import org.hisp.dhis.user.CurrentUserUtil;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserService;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service("org.hisp.dhis.tracker.export.trackedentity.TrackedEntityChangeLogService")
+@Transactional(readOnly = true)
+@RequiredArgsConstructor
+public class DefaultTrackedEntityChangeLogService implements TrackedEntityChangeLogService {
+
+  private final TrackedEntityService trackedEntityService;
+
+  private final ProgramService programService;
+
+  private final TrackedEntityProgramOwnerService trackedEntityProgramOwnerService;
+
+  private final TrackedEntityAttributeService trackedEntityAttributeService;
+
+  private final TrackerAccessManager trackerAccessManager;
+
+  private final UserService userService;
+
+  private final JdbcTrackedEntityChangeLogStore jdbcTrackedEntityChangeLogStore;
+
+  @Override
+  public Page<TrackedEntityChangeLog> getTrackedEntityChangeLog(
+      UID trackedEntityUid, UID programUid, PageParams pageParams) throws NotFoundException {
+    TrackedEntity trackedEntity =
+        trackedEntityService.getTrackedEntity(trackedEntityUid.getValue());
+    if (trackedEntity == null) {
+      throw new NotFoundException(TrackedEntity.class, trackedEntityUid.getValue());
+    }
+
+    User currentUser = userService.getUserByUsername(CurrentUserUtil.getCurrentUsername());
+    Set<String> trackedEntityAttributes = Collections.emptySet();
+    if (programUid != null) {
+      Program program = validateProgram(programUid.getValue());
+      validateOwnership(currentUser, trackedEntity, program);
+    } else {
+      validateTrackedEntity(trackedEntityUid, currentUser, trackedEntity);
+      trackedEntityAttributes = validateTrackedEntityAttributes(trackedEntity);
+    }
+
+    return jdbcTrackedEntityChangeLogStore.getTrackedEntityChangeLog(
+        trackedEntityUid, trackedEntityAttributes, pageParams);
+  }
+
+  private Program validateProgram(String programUid) throws NotFoundException {
+    Program program = programService.getProgram(programUid);
+    if (program == null) {
+      throw new NotFoundException(Program.class, programUid);
+    }
+
+    return program;
+  }
+
+  private void validateOwnership(User currentUser, TrackedEntity trackedEntity, Program program)
+      throws NotFoundException {
+    if (!trackerAccessManager
+        .canRead(currentUser, trackedEntity, program, currentUser.isSuper())
```

Comment:
I don't think it's wrong because the super user will likely have access to all org units in the system anyway. But if we don't skip we are doing an unnecessary validation. Also, we agreed with markus and enrico that when it comes to a super user, we would only validate its sharing settings.